File: 프로그래머스/3/49189. 가장 먼 노드/가장 먼 노드.py
from collections import deque

def solution(n, edge):
    adj = [[] for _ in range(n+1)]
    for s, e in edge:
        adj[s].append(e)
        adj[e].append(s)
    
    return bfs(n, adj)

def bfs(n, adj):
    queue = deque([1])
    visited = [0] * (n+1)
    visited[1] = 1
    
    while queue:
        now = queue.popleft()
        
        for next in adj[now]:
            if visited[next] == 0:
                queue.append(next)
                visited[next] = visited[now] + 1
    # print(visited)
    answer = 0
    max_v = 0
    for i in range(1, n+1):
        if visited[i] > max_v:
            max_v = visited[i]
            answer = 1
        elif visited[i] == max_v:
            answer += 1
    return answer
        

File: 프로그래머스/3/49189. 가장 먼 노드/가장 먼 노드.js
function solution(n, edge) {
    var answer = 0;
    const adj = Array.from({length:n+1}, () => Array(n+1).fill(0));
    for(const [u, v] of edge){
        adj[u][v] = 1;
        adj[v][u] = 1;
    }
    
    let farest = 0;
    let cnt = 0;
    
    const visited = Array.from({length:n+1}, () => [false, 0]);
    
    const q = [];
    let head = 0;
    
    q.push(1);
    visited[1][0] = true, visited[1][1] = 0;
    console.log()
    while(head < q.length){
        const now = q[head++];
        
        for(let next = 1; next<n+1; next++){
            if (adj[now][next] === 0){
                continue;
            }
            if (visited[next][0] === true){
                continue;
            }
            visited[next][0] = true;
            visited[next][1] = visited[now][1] + 1;
            if (visited[next][1] > farest){
                farest = visited[next][1];
                cnt = 1;
            }
            else if (visited[next][1] === farest){
                cnt++;
            }
            q.push(next);
        }
    }
    
    // console.log(visited);
    return cnt;
}